File: java/src/br/unifor/facilitador/dominio/Consulta.java
package br.unifor.facilitador.dominio;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

/**
 * SRP: esta classe so cuida do estado de uma consulta e das regras que dependem
 * apenas dele. Nao salva em banco, nao notifica ninguem, nao chama IA.
 */
public class Consulta {

    private final UUID id = UUID.randomUUID();
    private final Paciente paciente;
    private final Medico medico;
    private LocalDateTime dataHora;
    private Modalidade modalidade;
    private StatusConsulta status = StatusConsulta.SOLICITADA;
    private String motivoCancelamento;
    private final List<Documento> anexos = new ArrayList<>();

    public Consulta(Paciente paciente, Medico medico, LocalDateTime dataHora, Modalidade modalidade) {
        this.paciente = paciente;
        this.medico = medico;
        this.dataHora = dataHora;
        this.modalidade = modalidade;
        paciente.registrarConsulta(this);
        medico.registrarConsulta(this);
    }

    public void confirmar() {
        exigirStatus(StatusConsulta.SOLICITADA, "confirmar");
        this.status = StatusConsulta.CONFIRMADA;
    }

    public void cancelar(String motivo) {
        if (status == StatusConsulta.REALIZADA) {
            throw new IllegalStateException("Consulta ja realizada nao pode ser cancelada.");
        }
        this.status = StatusConsulta.CANCELADA;
        this.motivoCancelamento = motivo;
    }

    public void remarcar(LocalDateTime nova) {
        if (status == StatusConsulta.CANCELADA || status == StatusConsulta.REALIZADA) {
            throw new IllegalStateException("So consulta ativa pode ser remarcada.");
        }
        this.dataHora = nova;
        this.status = StatusConsulta.SOLICITADA;
    }

    public void anexar(Documento d) {
        anexos.add(d);
        paciente.getProntuario().anexar(d);
    }

    public List<Documento> getAnexos() {
        return Collections.unmodifiableList(anexos);
    }

    private void exigirStatus(StatusConsulta esperado, String acao) {
        if (this.status != esperado) {
            throw new IllegalStateException("Nao e possivel " + acao + " uma consulta " + status);
        }
    }

    public UUID getId() { return id; }
    public Paciente getPaciente() { return paciente; }
    public Medico getMedico() { return medico; }
    public LocalDateTime getDataHora() { return dataHora; }
    public Modalidade getModalidade() { return modalidade; }
    public StatusConsulta getStatus() { return status; }
    public String getMotivoCancelamento() { return motivoCancelamento; }

    @Override
    public String toString() {
        return "Consulta[" + dataHora + ", " + modalidade + ", " + status + "]";
    }
}

File: javascript/src/dominio/Consulta.js
import { randomUUID } from 'node:crypto';
import { StatusConsulta } from './StatusConsulta.js';

/**
 * SRP: esta classe so cuida do estado de uma consulta e das regras que dependem
 * apenas dele. Nao salva em banco, nao notifica ninguem, nao chama IA.
 */
export class Consulta {
  #id = randomUUID();
  #paciente;
  #medico;
  #dataHora;
  #modalidade;
  #status = StatusConsulta.SOLICITADA;
  #motivoCancelamento = null;
  #anexos = [];

  constructor(paciente, medico, dataHora, modalidade) {
    this.#paciente = paciente;
    this.#medico = medico;
    this.#dataHora = dataHora;
    this.#modalidade = modalidade;
    paciente.registrarConsulta(this);
    medico.registrarConsulta(this);
  }

  confirmar() {
    if (this.#status !== StatusConsulta.SOLICITADA) {
      throw new Error(`Nao e possivel confirmar uma consulta ${this.#status}`);
    }
    this.#status = StatusConsulta.CONFIRMADA;
  }

  cancelar(motivo) {
    if (this.#status === StatusConsulta.REALIZADA) {
      throw new Error('Consulta ja realizada nao pode ser cancelada.');
    }
    this.#status = StatusConsulta.CANCELADA;
    this.#motivoCancelamento = motivo;
  }

  remarcar(nova) {
    if (this.#status === StatusConsulta.CANCELADA || this.#status === StatusConsulta.REALIZADA) {
      throw new Error('So consulta ativa pode ser remarcada.');
    }
    this.#dataHora = nova;
    this.#status = StatusConsulta.SOLICITADA;
  }

  anexar(documento) {
    this.#anexos.push(documento);
    this.#paciente.prontuario.anexar(documento);
  }

  get anexos() { return Object.freeze([...this.#anexos]); }
  get id() { return this.#id; }
  get paciente() { return this.#paciente; }
  get medico() { return this.#medico; }
  get dataHora() { return this.#dataHora; }
  get modalidade() { return this.#modalidade; }
  get status() { return this.#status; }
  get motivoCancelamento() { return this.#motivoCancelamento; }

  toString() {
    return `Consulta[${this.#dataHora.toISOString()}, ${this.#modalidade}, ${this.#status}]`;
  }
}
